Running compute job with user defined parameters (#1113)
oceanprotocol/docs: building-with-ocean/compute-to-data/user-defined-parameters.md

---
description: >-
  Learn how to define and use custom parameters while downloading assets or
  using dataset in Compute-to-data environment
---

# User defined parameters

### Overview

Ocean Protocol allows dataset buyers to provide custom parameters that can be used to fetch the downloaded data in a specific format, download a different type of data or pass some additional input to the algorithms in the Compute-to-Data job.

There 2 are types of parameters that asset publishers can support:

* User defined parameters
* Algorithm custom parameters

### Publish a dataset that uses custom parameters

> The dataset publisher can support these parameters to allow filtering, querying on the published dataset. The additional parameters that facilitate this are called `User defined parameters`. The Provider combines the original asset URL and the entered parameter values into a new URL and then streams the response from the modified URL back to the buyer.

Reviewer: ```suggestion
The dataset publisher can support these parameters to allow filtering or querying of the published dataset. The additional parameters that facilitate this are called `User defined parameters`. The Provider combines the original asset URL and the entered parameter values into a new URL and then streams the response from the modified URL back to the buyer.
```

#### Use case for user defined parameters

> For example, if the publisher has published an URL `https://example.com` which serves large size historical weather data from all over the world. If the publisher wants to allow buyers to filter the data based on location, type of data, etc., it is possible to do so using user defined parameters.

Reviewer: ```suggestion
For example, if the publisher has published an URL `https://example.com` which serves large size historical weather data from all over the world, the publisher could allow buyers to filter the data based on location, type of data, etc. It is possible to do this using user defined parameters.
```

Suppose the publisher defines the following 2 parameters:

* `location`: A string indicating region code
* `type`: A string indicating the type of weather data. It can be temperature/humidity/pressure.

> Suppose the buyer wants to download the temperature data in the region code `XYZ`. While downloading the data buyer enters the desired parameter values using ocean.py or ocean.js.

Reviewer: ```suggestion
Suppose the buyer wants to download the temperature data in the region code `XYZ`. While downloading the data, the buyer enters the desired parameter values using ocean.py or ocean.js.
```

Reviewer: Can this be done with ocean.js?

Author: Not as of now. But, I think it will be soon supported in ocean.js as well.

The following steps will specify how the publisher can support additional parameters.

#### Step 1: Create a service

The below python script exposes a REST endpoint that takes two parameters, namely: `location` and `type`. Let's assume that the dataset publisher hosts the service at domain `example.com` along with HTTPS support. The publisher must ensure that the URL is accessible to Provider.

The code snippet is only for demo purposes and not for production use.

```python
from flask import Flask, request
def get_data(data_type: str, location: str):
    '''
        Add some business logic here to get
        the required data with given parameters
    '''
    return {}
@app.route('/', methods=['GET'])
def serve_content():
    args = request.args
    data_type = args.get('type')
    location = args.get('location')
    result = get_data(data_type, location)
    return result
```

#### Step 2: Publish dataset asset with compute service

The publisher now must provide the file URL as `https://example.com` while publishing the asset, as shown in the below image.

![](../../.gitbook/assets/compute-to-data-parameters-publish-dataset.png)

> To view a complete tutorial on publishing asset using Ocean Marketplace click [here](../../using-ocean-market/marketplace-publish-data-asset.md).

Reviewer: ```suggestion
For a complete tutorial on publishing asset using Ocean Marketplace read [our guide on publishing with Ocean Market](../../using-ocean-market/marketplace-publish-data-asset.md).
```

Reviewer: In general, let's avoid saying "click here". It's nicer if the link actually describes the content.

The following steps will specify how the algorithm publisher can support additional algorithm custom parameters.

#### Step 1: Create an algorithm

The code snippet is only for demo purposes and not for production use.

```python
def run_algorithm(i: int):
    pass
def read_algorithm_custom_input():
    parameters_file = os.path.join(os.sep, "data", "inputs", "algoCustomData.json")
    with open(parameters_file, "r") as file:
        return json.load(file)
algorithm_inputs = read_algorithm_custom_input()
iterations = algorithm_inputs["iterations"]
for i in range(iterations):
    # Run some machine learning algorithm
    print(f"Running iteration {i}")
    result = run_algorithm(i)
output_dir = os.path.join(os.sep, "data", "outputs")
with open(os.path.join(output_dir, "result"), "w") as f:
    f.write(result)
```

#### Step 2: Publish algorithm asset

The publisher now must provide the file URL as `https://example.org` while publishing the algorithm asset, as shown in the below image.

![](../../.gitbook/assets/compute-to-data-parameters-publish-algorithm.png)

> To view a complete tutorial on publishing asset using Ocean Marketplace click [here](../../using-ocean-market/marketplace-publish-data-asset.md).

Reviewer: ```suggestion
For a complete tutorial on publishing asset using Ocean Marketplace read [our guide on publishing with Ocean Market](../../using-ocean-market/marketplace-publish-data-asset.md).
```

### Starting compute job with custom parameters

> After the publishing, assume that a buyer wants to run the with certain parameters algorithm on selected dataset. The code snippet below shows how the buyer can start the compute job with custom parameter values. Searching dataset, allowing algorithm to run the dataset, buying datatokens, setting up ocean.py/ocean.js is out of scope of this tutorial.

Reviewer: ```suggestion
In this example, the buyer wants to run the algorithm with certain parameters on a selected dataset. The code snippet below shows how the buyer can start the compute job with custom parameter values. Searching the dataset, allowing the algorithm to run on the dataset, buying datatokens, setting up ocean.py/ocean.js is out of scope of this tutorial.
```

Reviewer: Rather than listing the things that are out of scope, it would be better to provide a list of links where they can read about how to do those things. 

e.g. "Before embarking on this tutorial you should familiarise yourself with how to [setup ocean.js/ocean.py](link), [buy datatokens](link), [run algorithms](link) etc..."
